### PR TITLE
reside-89: blog post about debugging CRAN memory issues

### PR DESCRIPTION
**Pull request by richfitz**

### CODE REVIEW

File: content/blog/debug-cran-additional.md [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+
+To trigger this, I used the [`wch1/r-debug`](https://hub.docker.com/r/wch1/r-debug) docker container as I've had success with this in the past.  There are also containers from the [rocker project](https://www.rocker-project.org/) that a similar approach will likely work with, such as [`rocker/r-devel-ubsan-clang`](https://hub.docker.com/r/rocker/r-devel-ubsan-clang).
+
+First, I started the container with
+
+```
+docker run -v $PWD:/src:ro -it --rm --security-opt seccomp=unconfined wch1/r-debug
+```
+
+then, in that container, I prepared the R library in that container using the R binary that was built with the undefined behaviour checkers enabled:
```
Comment (giovannic) [inline content/blog/debug-cran-additional.md]: in that container twice

File: content/blog/debug-cran-additional.md [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+  size_t idx0 = 0;
+  if (n > 0) {
+    const double
+      t0 = ((double*) ring_buffer_tail(obj->history))[idx_t],
+      t1 = ((double*) ring_buffer_tail_offset(obj->history, n - 1))[idx_t];
+    idx0 = min_size((t - t0) / (t1 - t0) / (n - 1), n - 1);
+  }
+```
+
+When triggered (the second assignment to `idx0`), we have (approximately) `t0 = 11`, `t1 = 12` and `t = 0.02` which falls outside of the range of times, so the expression `(t - t0) / (t1 - t0) / (n - 1)` is negative and that's the undefined behaviour.
```
Comment (giovannic) [inline content/blog/debug-cran-additional.md]: I can accept this but it's hard to follow when you're not sure what the code is meant to be doing, let alone what t0 and t1 are.

**Review by giovannic — COMMENTED** [review]
Super interesting post!

Wrt the narrative, I'm confused about how you knew about the second two issues. Were they also flagged by CRAN or were you simply going the extra mile to make it airtight? That made the jump from the "Undefined behaviour" section to the "Valgrind" and "rchk" sections a bit stark.

Cool!

File: content/blog/debug-cran-additional.md (hunk 2 above)
Comment (richfitz) [reply content/blog/debug-cran-additional.md]: thanks - @weshinsley was confused here too so I have expanded a bit

**Comment by richfitz** [comment]
Thanks @giovannic - very helpful

File: content/blog/debug-cran-additional.md [hunk 3]
```diff
@@ -0,0 +1,384 @@
+---
+title: "Debugging and Fixing CRAN's 'Additional Checks' errors"
+date: 2020-01-16
+tags:
+ - dde
+ - R
+---
+
+R packages that are published on CRAN are tested every night on a variety of platforms and on the development version of R to ensure that they continue to work.  In addition, packages that contain compiled code (C, C++ or Fortran) are put through a raft of additional checks to ensure that the compiled code will not cause R to crash.  Once an issue is found, the package maintainer gets an email and usually a fairly short window to fix the package before it is removed from CRAN.  However, replicating the error locally can require installation of all sorts of esoteric tools (and a copy of the development version of R from source) and it's not always obvious to start.
+
```
Comment (weshinsley) [inline content/blog/debug-cran-additional.md:11]: possibly colloquialism - did you mean "where" to start or "how" to start - or perhaps it's fine as it is.

File: content/blog/debug-cran-additional.md [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+This blog post documents the process I used in clearing three issues from our [`dde`](https://mrc-ide.github.io/dde/) package (which implements a simple solver for delay differential equations - the astute blog reader may recognise it from [previous debugging efforts](https://reside-ic.github.io/blog/debugging-at-the-edge-of-reason/)).
+
+Package authors (or at least I) typically find out about these problems when getting an email from CRAN but they may have been live for some time and are listed as an "Additional issues" section on a packages [check page](https://cran.r-project.org/web/checks/check_results_dde.html) - however, this is shown only when there is a problem!
+
```
Comment (weshinsley) [inline content/blog/debug-cran-additional.md:15]: packages -> either package or package's
and (optionally) - perhaps listed *in* an "Additional issues" section.

File: content/blog/debug-cran-additional.md [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+  if (n > 0) {
+    const double
+      t0 = ((double*) ring_buffer_tail(obj->history))[idx_t],
+      t1 = ((double*) ring_buffer_tail_offset(obj->history, n - 1))[idx_t];
+    idx0 = min_size((t - t0) / (t1 - t0) / (n - 1), n - 1);
+  }
+```
+
+This is an optimisation to seed a binary search for a value close to `t` in an array of values from `t0` and `t1` - we assume that the values between `t0` and `t1` are roughly evenly spaced and linearly interpolate between them to get a likely enough index for `t`, which we store as `idx0`.
+
```
Comment (weshinsley) [inline content/blog/debug-cran-additional.md:120]: possibly from `t0` ~and~ *to* `t1` ?

This para is helpful though, in understanding why the negative might reasonably happen, and why it doesn't fundamentally matter, but rather eventually means just a not-so-successful optimisation.

Comment (weshinsley) [reply content/blog/debug-cran-additional.md:120]: Train of thought - don't know whether this is more detail than wanted...

(1) You could clarify: `n` is of type `size_t`

(2) `(t-t0) / (t1-t0) / (n-1)` evaluates left to right, hence ends up being an integer division, with resulting type `size_t`, (not `double` - I was briefly confused wondering why gcc didn't squawk more about this, until I looked up what `n` was))

(3) Could mention `min_size` takes two `size_t` args - because it's in the `min_size` function call itself that the undefined behaviour warning is triggered when binding the first arg?

(4) So what I think's actually happening (ie, how this made no difference), is that when min_size is called, the negative number gets cast to a size_t (with the warning), and wraps round into a relatively large (unsigned) positive integer, hence min_size always returns the smaller argument, which is (n-1).

(5) Hence, if you wanted *exactly* the same behaviour in your fixed code, you could say, `else idx = n-1;` after your `<0` comparison - but it's probably not worth it, and takes more explaining!

File: content/blog/debug-cran-additional.md [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+    }
+  }
+```
+
+With this fix in place, the UBSAN checks pass without incident.  See [`8c384bb`](https://github.com/mrc-ide/dde/commit/8c384bb8bcc0f2513fa955ce4951b65cefb0dfa1) for details.
+
+## Valgrind
+
+[Valgrind](https://valgrind.org/) finds memory errors and is one of my favourite tools for working out why something crashes.  My (probably grossly oversimplified) understanding is that it runs a program with a layer that checks all memory accesses for correctness (alignment, bounds etc).  Unsurprisingly this makes things very slow to run.  The information on the CRAN page was:
+
```
Comment (weshinsley) [inline content/blog/debug-cran-additional.md:142]: That's exactly what I think valgrind does too... :-)

File: content/blog/debug-cran-additional.md [hunk 7]
```diff
@@ -0,0 +1,384 @@
...
+```
+
+which is what [is documented to work](https://www.stats.ox.ac.uk/pub/bdr/memtests/README.txt) did not yield the error.  But with a bit of poking based on the `Command:` line in the above valgrind output I got some success with:
+
+```
+(cd dde/tests && RDvalgrind -d valgrind -f testthat.R --no-readline --vanilla)
+```
+
+which shows the invalid read, along side some less interesting output.  Using the `LocationReporter` again was better still:
+
```
Comment (weshinsley) [inline content/blog/debug-cran-additional.md:216]: I think alongside is one word

File: content/blog/debug-cran-additional.md [hunk 8]
```diff
@@ -0,0 +1,384 @@
...
+```
+    double t0 = obj->sign * times[0];
+    while (obj->sign * tcrit[obj->tcrit_idx] <= t0 &&
+           obj->tcrit_idx < n_tcrit) {
+      obj->tcrit_idx++;
+    }
+```
+
+which is a nasty enough bit of book-keeping.  But the problem is that the while condition's two clauses are in the wrong order - when `obj->tcrit_idx < n_tcrit` is `false` we *really* should not be looking `tcrit[obj->tcrit_idx]` because it is out of range, which is precisely what the "invalid read" error valgrind reported was.  The fixed code looks like
+
```
Comment (weshinsley) [inline content/blog/debug-cran-additional.md:295]: looking *at*, or looking *up* perhaps?

File: content/blog/debug-cran-additional.md [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+First, I started the container with
+
+```
+docker run -v $PWD:/src:ro -it --rm --security-opt seccomp=unconfined wch1/r-debug
+```
+
+then, in that container, I prepared the R library using the R binary that was built with the undefined behaviour checkers enabled:
+
+```
+RDcsan -e 'install.packages(c("deSolve", "knitr", "ring", "microbenchmark", "rmarkdown", "testthat", "devtools", "roxygen2"))'
```
Comment (r-ash) [inline content/blog/debug-cran-additional.md:58]: Was a bit confused about what `RDcsan` was on first reading. Maybe worth a small note? Or linking to the readme for the repo https://github.com/wch/r-debug

**Review by r-ash — COMMENTED** [review]
One small comment
